### PR TITLE
added json formatter  for  unittest module

### PR DESCRIPTION
**Pull request by ynfle**
Added the JSON formatter for the unittest module to be able to output the desired format to integrate with Exercism

### CODE REVIEW

File: src/unittest_json.nim [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    stream*: Stream
+    testErrors*: seq[string]
+    testStackTrace*: string
+    result*: ResultJson
+    ranFirstTest*: bool
+
+  JsonTestStatus* = enum
+    PASS,
+    FAIL,
+    ERROR
```
Comment (ee7) [inline src/unittest_json.nim:18]:
```suggestion
    PASS = "pass",
    FAIL = "fail",
    ERROR = "error"
```
This changes the output values in the JSON file to be lowercase, as specified here https://github.com/exercism/automated-tests/blob/master/docs/interface.md#statuses

File: src/unittest_json.nim [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    of ERROR:
+      # TODO: Clarify when status is error
+      message*: string
+    of PASS, FAIL:
+      discard
+
+
+proc newJsonOutputFormatter*(stream: Stream): <//>JsonOutputFormatter =
+  ## Creates a formatter that writes report to the specified stream in
+  ## JUnit format.
```
Comment (ee7) [inline src/unittest_json.nim:41]:
```suggestion
  ## JSON format.
```

File: src/unittest_json.nim [hunk 3]
```diff
@@ -0,0 +1,122 @@
+##
+## An extension of the ``unittest.nim`` library to output json for Exercism v3
+##
+
+import json, unittest, streams, strformat, strutils
```
Comment (ee7) [inline src/unittest_json.nim:5]:
```suggestion
import json, streams, strformat, unittest
```
Remove unused `strutils` import and order alphabetically.

File: src/unittest_json.nim [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  ResultJson* = ref object
+    tests*: seq[JsonTestResult]
+    case status*: JsonTestStatus
+    of ERROR:
+      # TODO: Clarify when status is error
+      message*: string
+    of PASS, FAIL:
+      discard
+
+
```
Comment (ee7) [inline src/unittest_json.nim:38]:
```suggestion
```

File: src/unittest_json.nim [hunk 5]
```diff
@@ -0,0 +1,122 @@
+##
+## An extension of the ``unittest.nim`` library to output json for Exercism v3
+##
```
Comment (ee7) [inline src/unittest_json.nim:3]:
```suggestion
## An extension of the ``unittest.nim`` library to output json for Exercism v3
```

File: src/unittest_json.nim [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        else: formatter.testErrors.len - 1
+
+      for errIdx in startIdx..endIdx:
+        if errs.len > 0:
+          errs.add("\n")
+        errs.add(formatter.testErrors[errIdx])
+
+    echo "Summary"
+    if formatter.testStackTrace.len > 0:
+      jsonTestResult = JsonTestResult(name: testResult.testName, status: ERROR, message: fmt"{failureMsg}\n{formatter.testStackTrace}")
```
Comment (ee7) [inline src/unittest_json.nim:107]:
```suggestion
      jsonTestResult = JsonTestResult(name: testResult.testName, status: ERROR,
          message: fmt"{failureMsg}\n{formatter.testStackTrace}")
```

File: src/unittest_json.nim [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+      echo "\tfailureMsg: ", failureMsg
+      echo "\ttestStackTrace: ", formatter.testStackTrace
+      if errs.len > 0:
+        echo "\terrs: ", errs
+        jsonTestResult.message.insert(errs & "\n")
+    else:
+      echo "\tfailureMsg: ", failureMsg
+      echo "\terrs: ", errs
+      jsonTestResult = JsonTestResult(name: testResult.testName, status: FAIL, message: fmt"{failureMsg}\n{errs}")
```
Comment (ee7) [inline src/unittest_json.nim:116]:
```suggestion
      jsonTestResult = JsonTestResult(name: testResult.testName, status: FAIL,
          message: fmt"{failureMsg}\n{errs}")
```

File: src/unittest_json.nim (hunk 6 above)
Comment (ynfle) [reply src/unittest_json.nim:107]: Should we move all three arguments to their own lines?

File: src/unittest_json.nim (hunk 7 above)
Comment (ynfle) [reply src/unittest_json.nim:116]: Same as https://github.com/exercism/nim-test-runner/pull/2/files#r422414019

File: src/unittest_json.nim (hunk 6 above)
Comment (ee7) [reply src/unittest_json.nim:104]: Sure.

File: src/unittest_json.nim [hunk 8]
```diff
@@ -0,0 +1,127 @@
+## An extension of the ``unittest.nim`` library to output json for Exercism v3
+
+import json, streams, strformat, unittest
+
+type
+  JsonOutputFormatter* = ref object of OutputFormatter
+    stream*: Stream
+    testErrors*: seq[string]
+    testStackTrace*: string
+    result*: ResultJson
```
Comment (ee7) [inline src/unittest_json.nim:10]:
```suggestion
  JsonOutputFormatter = ref object of OutputFormatter
    stream: Stream
    testErrors: seq[string]
    testStackTrace: string
    result: ResultJson
```

File: src/unittest_json.nim [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    ranFirstTest*: bool
```
Comment (ee7) [inline src/unittest_json.nim:11]:
```suggestion
```

File: src/unittest_json.nim [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+  JsonTestStatus* = enum
```
Comment (ee7) [inline src/unittest_json.nim:13]:
```suggestion
  JsonTestStatus = enum
```

File: src/unittest_json.nim [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    PASS = "pass",
+    FAIL = "fail",
+    ERROR = "error"
+
+  JsonTestResult* = object
+    name*: string
+    output*: string
+    case status*: JsonTestStatus
+    of FAIL, ERROR:
+      message*: string
```
Comment (ee7) [inline src/unittest_json.nim:23]:
```suggestion
  JsonTestResult = object
    name: string
    output: string
    case status: JsonTestStatus
    of FAIL, ERROR:
      message: string
```

File: src/unittest_json.nim [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+    of PASS: 
+      discard
+
+  ResultJson* = ref object
+    tests*: seq[JsonTestResult]
+    case status*: JsonTestStatus
```
Comment (ee7) [inline src/unittest_json.nim:29]:
```suggestion
  ResultJson = ref object
    tests: seq[JsonTestResult]
    case status: JsonTestStatus
```

File: src/unittest_json.nim [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    of ERROR:
+      # TODO: Clarify when status is error
```
Comment (ee7) [inline src/unittest_json.nim:31]:
```suggestion
```

File: src/unittest_json.nim [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+      message*: string
```
Comment (ee7) [inline src/unittest_json.nim:32]:
```suggestion
      message: string
```

File: src/unittest_json.nim [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+  ## Creates a formatter that writes report to the specified stream in
+  ## JSON format.
+  ## The ``stream`` is NOT closed automatically when the test are finished,
+  ## because the formatter has no way to know when all tests are finished.
+  ## You should invoke formatter.close() to finalize the report.
+  result = JsonOutputFormatter(
+    result: ResultJson(tests: @[], status: PASS),
+    stream: stream,
+    testErrors: @[],
+    testStackTrace: "",
```
Comment (ee7) [inline src/unittest_json.nim:46]:
```suggestion
    stream: stream,
    testErrors: @[],
    testStackTrace: "",
    result: ResultJson(tests: @[], status: PASS)
```
Order closer to `newJUnitOutputFormatter`.

File: src/unittest_json.nim [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    ranFirstTest: false
```
Comment (ee7) [inline src/unittest_json.nim:47]:
```suggestion
```

File: src/unittest_json.nim [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+  )
+  echo "original result: ", %result.result
```
Comment (ee7) [inline src/unittest_json.nim:49]:
```suggestion
```

File: src/unittest_json.nim [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+
+proc close*(formatter: JsonOutputFormatter) =
+  ## Completes the report and closes the underlying stream.
+  # echo "errors: ", formatter.testErrors
```
Comment (ee7) [inline src/unittest_json.nim:53]:
```suggestion
```

File: src/unittest_json.nim [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+  formatter.stream.write($(%formatter.result))
+  formatter.stream.close()
+
+method suiteStarted*(formatter: JsonOutputFormatter, suiteName: string) =
```
Comment (ee7) [inline src/unittest_json.nim:57]:
```suggestion
method suiteStarted(formatter: JsonOutputFormatter, suiteName: string) =
```

File: src/unittest_json.nim [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+  discard
+
+method testStarted*(formatter: JsonOutputFormatter, testName: string) =
```
Comment (ee7) [inline src/unittest_json.nim:60]:
```suggestion
method testStarted(formatter: JsonOutputFormatter, testName: string) =
```

File: src/unittest_json.nim [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+  formatter.testErrors.setLen(0)
+  formatter.testStackTrace.setLen(0)
+
+method failureOccurred*(formatter: JsonOutputFormatter,
+                        checkpoints: seq[string], stackTrace: string) =
```
Comment (ee7) [inline src/unittest_json.nim:65]:
```suggestion
method failureOccurred(formatter: JsonOutputFormatter,
                       checkpoints: seq[string], stackTrace: string) =
```

File: src/unittest_json.nim [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+  ## ``stackTrace`` is provided only if the failure occurred due to an
+  ## exception. ``checkpoints`` is never ``nil``.
+  echo "Test Failed"
+  echo "\tcheckpoints: ", checkpoints
+  echo "\tstackTrace: ", stackTrace
```
Comment (ee7) [inline src/unittest_json.nim:70]:
```suggestion
```

File: src/unittest_json.nim [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+  formatter.testErrors.add(checkpoints)
+  if stackTrace.len > 0:
+    formatter.testStackTrace = stackTrace
+
+method testEnded*(formatter: JsonOutputFormatter, testResult: TestResult) =
```
Comment (ee7) [inline src/unittest_json.nim:75]:
```suggestion
method testEnded(formatter: JsonOutputFormatter, testResult: TestResult) =
```

File: src/unittest_json.nim [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+    jsonTestResult = JsonTestResult(name: testResult.testName, status: PASS)
+  of TestStatus.SKIPPED:
+    discard
+  of TestStatus.FAILED:
+    let failureMsg = if formatter.testStackTrace.len > 0 and
+                        formatter.testErrors.len > 0:
+                       formatter.testErrors[^1]
+                     elif formatter.testErrors.len > 0:
+                       formatter.testErrors[0]
+                     else: "The test failed without outputting an error"
```
Comment (ee7) [inline src/unittest_json.nim:88]:
```suggestion
    let failureMsg =
      if formatter.testStackTrace.len > 0 and formatter.testErrors.len > 0:
        formatter.testErrors[^1]
      elif formatter.testErrors.len > 0:
        formatter.testErrors[0]
      else:
        "The test failed without outputting an error"
```

File: src/unittest_json.nim [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+
+    var errs = ""
+    if formatter.testErrors.len > 1:
+      var startIdx = if formatter.testStackTrace.len > 0: 0 else: 1
```
Comment (ee7) [inline src/unittest_json.nim:92]:
```suggestion
      let startIdx = if formatter.testStackTrace.len > 0: 0 else: 1
```

File: src/unittest_json.nim [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+      var endIdx = if formatter.testStackTrace.len > 0:
+          formatter.testErrors.len - 2
+        else: formatter.testErrors.len - 1
```
Comment (ee7) [inline src/unittest_json.nim:95]:
```suggestion
      let endIdx =
        if formatter.testStackTrace.len > 0:
          formatter.testErrors.len - 2
        else:
          formatter.testErrors.len - 1
```
Change `var` to `let`

File: src/unittest_json.nim [hunk 27]
```diff
@@ -0,0 +1,127 @@
...
+
+      for errIdx in startIdx..endIdx:
+        if errs.len > 0:
+          errs.add("\n")
+        errs.add(formatter.testErrors[errIdx])
+
+    echo "Summary"
```
Comment (ee7) [inline src/unittest_json.nim:102]:
```suggestion
```

File: src/unittest_json.nim [hunk 28]
```diff
@@ -0,0 +1,127 @@
...
+    if formatter.testStackTrace.len > 0:
+      jsonTestResult = JsonTestResult(
+        name: testResult.testName,
+        status: ERROR,
```
Comment (ee7) [inline src/unittest_json.nim:106]:
```suggestion
        status: FAIL,
```
I think it should be `FAIL` here? To my understanding, the following implementation of `hello-world` should produce a test `"status: "fail"`, not `status: "error"`.
```
proc hello*: string =
  raise newException(ValueError, "hi")
  "Hello, World!"
```
"If a student's code throws an exception and that was not what the test expected, it should just be a regular test failure."

File: src/unittest_json.nim [hunk 29]
```diff
@@ -0,0 +1,127 @@
...
+        message: fmt"{failureMsg}\n{formatter.testStackTrace}"
+      )
+      echo "\tfailureMsg: ", failureMsg
+      echo "\ttestStackTrace: ", formatter.testStackTrace
```
Comment (ee7) [inline src/unittest_json.nim:110]:
```suggestion
```

File: src/unittest_json.nim [hunk 30]
```diff
@@ -0,0 +1,127 @@
...
+      if errs.len > 0:
+        echo "\terrs: ", errs
```
Comment (ee7) [inline src/unittest_json.nim:112]:
```suggestion
```

File: src/unittest_json.nim [hunk 31]
```diff
@@ -0,0 +1,127 @@
...
+        jsonTestResult.message.insert(errs & "\n")
+    else:
+      echo "\tfailureMsg: ", failureMsg
+      echo "\terrs: ", errs
```
Comment (ee7) [inline src/unittest_json.nim:116]:
```suggestion
```

File: src/unittest_json.nim [hunk 32]
```diff
@@ -0,0 +1,127 @@
...
+      jsonTestResult = JsonTestResult(
+        name: testResult.testName,
+        status: FAIL,
+        message: fmt"{failureMsg}\n{errs}"
+      )
+    formatter.result = ResultJson(status: FAIL, tests: formatter.result.tests)
+
+  formatter.result.tests.add(jsonTestResult)
+
+method suiteEnded*(formatter: JsonOutputFormatter) =
```
Comment (ee7) [inline src/unittest_json.nim:126]:
```suggestion
method suiteEnded(formatter: JsonOutputFormatter) =
```

File: src/unittest_json.nim (hunk 28 above)
Comment (ynfle) [reply src/unittest_json.nim:106]: Ok. Makes sense. What is the purpose of the error status at the test level?

Comment (ee7) [reply src/unittest_json.nim:100]: I think it's more for the non-compiled languages.

I believe that, for us, the expected JSON output for

```
proc hello*: string = 
  1
```

with the test

```
import unittest
import hello_world

# version 1.1.0

suite "Hello World":
  test "say hi":
    check hello() == "Hello, World!"
```

is something like

```
{
  "status": "error",
  "message": "hello_world.nim(2, 3) Error: type mismatch: got <int literal(1)> but expected 'string'",
  "tests": []
}
```

That is, the value of the `tests` key is the empty array - there is no per-test `status: "error"`.

File: src/unittest_json.nim [hunk 33]
```diff
@@ -0,0 +1,116 @@
...
+    of PASS: 
+      discard
+
+  ResultJson = ref object
+    tests: seq[JsonTestResult]
+    case status: JsonTestStatus
+    of ERROR:
+      message: string
+    of PASS, FAIL:
+      discard
```
Comment (ee7) [inline src/unittest_json.nim:32]:
```suggestion
  JsonTestStatus = enum
    PASS = "pass",
    FAIL = "fail",
    ERROR = "error"

  JsonTestResult = object
    name: string
    output: string
    case status: JsonTestStatus
    of FAIL, ERROR:
      message: string
    of PASS:
      discard

  ResultJson = ref object
    tests: seq[JsonTestResult]
    case status: JsonTestStatus
    of ERROR:
      message: string
    of PASS, FAIL:
      discard

  JsonOutputFormatter = ref object of OutputFormatter
    stream: Stream
    testErrors: seq[string]
    testStackTrace: string
    result: ResultJson
```

**Pull request merged**